src/linker/ref/Linker.Steps/MarkHandlerDispatcher.cs

```diff
@@ -0,0 +1,19 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace Mono.Linker.Steps
+{
+	public abstract class MarkHandlerDispatcher : IMarkHandler
+	{
+		protected MarkHandlerDispatcher () => throw null;
+
+		protected MarkHandlerDispatcher (IEnumerable<ISubStep> subSteps) => throw null;
+
+		public void Add (ISubStep substep) => throw null;
```

Comment:
Hmm, do we have any use case for this? It seems to me this is a more complicated version of `MarkContext` methods.

Reply:
This is a convenience helper for steps that want to scan for all members (not just marked members) in marked assemblies. It is basically a copy of SubStepsDispatcher that works only for marked assemblies, to make it easier for existing substeps to transition - this way they don't have to re-write the code that scans all members in an assembly.

It's really only used for ApplyPreserveAttribute and PreserveExportedTypes ([here](https://github.com/xamarin/xamarin-android/compare/master...sbomer:illinkMarkHandler#diff-8484978825e4cf48cbaac59d46123e249f0ba65383625795dbc9ef1e46636dbcR45 )) - so maybe it's not worth keeping around.

Reply:
I'd rather see the public interface simple. Looking quickly at the XA code is seems that both are not needed

ApplyPreserveAttribute - that's normal step that should just run before MarkStep, it does not need any of the changes
PreserveExportedTypes - should run only when a type is marked

Reply:
If ApplyPreserveAttribute runs before MarkStep it won't see the reference closure and it won't run for dynamically referenced assemblies - is it safe to look for the attribute only in root assemblies? (Do we not care about the link-all case?)

My understanding of ExportAttribute (from https://docs.microsoft.com/en-us/xamarin/android/platform/java-integration/working-with-jni#exportattribute-and-exportfieldattribute) was that it is used on members which are reachable from java interop, but maybe that link doesn't ell the whole story. What is it that marks the required types?

Reply:
Yep, at a closer look, they need to run once assembly is marked. Thinking about it more it's probably fine to have both approaches available.

Few suggestions though.
- I'd rename this type to be `MarkSubStepsDispatcher` to make it clear this type should be used to batch all custom mark substeps.
- Ensure `MarkHandlerDispatcher.Initialize` is always called (the user type can implement explicitly IMarkHandler again).

Reply:
For the second suggestion - not sure if this is what you meant, but I changed Initialize to be a virtual method so that the user type can call `base.Initialize`.